File: day03/src/main/java/de/havox_design/aoc2017/day03/Point.java
package de.havox_design.aoc2017.day03;

import static de.havox_design.aoc2017.day03.Direction.*;

import java.util.Set;

public record Point(int x, int y) {
    public Point delta(int dx, int dy) {
        return new Point(this.x + dx, this.y + dy);
    }

    public Set<Point> neighbors() {
        return Set.of(
                this.delta(DOWN.getDx(), DOWN.getDy()),
                this.delta(LEFT.getDx(), LEFT.getDy()),
                this.delta(RIGHT.getDx(), RIGHT.getDy()),
                this.delta(UP.getDx(), UP.getDy())
        );
    }
}

File: day14/src/main/scala/de/havox_design/aoc2017/day14/Point.scala
package de.havox_design.aoc2017.day14

val orthogonal = Set((-1, 0), (1, 0), (0, -1), (0, 1))

case class Point(x: Int, y: Int):
  private def delta(dx: Int, dy: Int): Point = Point(x + dx, y + dy)
  def neighbours: Set[Point] = orthogonal.map(delta)
